### PR TITLE
[opt](Nereids)Join cluster connectivity

### PR DESCRIPTION
**Pull request by englefly**
## Proposed changes
in a join cluster graph, if a node has higher connectivity, it is more likely to be reduced by runtime filters, and it is more likely to produce effective runtime filters.
Thus, we prefer to put the node with higher connectivity on the join right side.

This pr is good for tpch q21 when col stats are not available. 

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by englefly** [comment]
run buildall

**Comment by englefly** [comment]
run buildall

**Comment by englefly** [comment]
run buildall

**Comment by englefly** [comment]
run buildall

**Review by starocean999 — DISMISSED** [review]

File: fe/fe-core/src/main/java/org/apache/doris/nereids/StatementContext.java [hunk 1]
```diff
@@ -88,6 +89,8 @@ public class StatementContext {
     private final Map<CTEId, LogicalPlan> rewrittenCteConsumer = new HashMap<>();
     private final Set<String> viewDdlSqlSet = Sets.newHashSet();
 
+    private final List<Expression> joinFilters = new ArrayList<>();
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/StatementContext.java:92]: add some comment to explain what it is use for

File: fe/fe-core/src/main/java/org/apache/doris/nereids/cost/CostModelV1.java [hunk 2]
```diff
@@ -310,6 +325,15 @@ public Cost visitPhysicalHashJoin(
         );
     }
 
+    private int computeConnectivity(
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/cost/CostModelV1.java:328]: add comment to explain it

File: fe/fe-core/src/main/java/org/apache/doris/nereids/cost/CostModelV1.java [hunk 3]
```diff
@@ -262,6 +265,18 @@ public Cost visitPhysicalHashJoin(
 
         double leftRowCount = probeStats.getRowCount();
         double rightRowCount = buildStats.getRowCount();
+        double ignorableDiffBetweenLeftAndRightThreshold = 10.0;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/cost/CostModelV1.java:268]: add comment to explain it

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/join/JoinCommute.java [hunk 4]
```diff
@@ -93,7 +93,7 @@ public static boolean check(SwapType swapType, LogicalJoin<GroupPlan, GroupPlan>
         if (swapType == SwapType.LEFT_ZIG_ZAG) {
             double leftRows = join.left().getGroup().getStatistics().getRowCount();
             double rightRows = join.right().getGroup().getStatistics().getRowCount();
-            return leftRows < rightRows && isZigZagJoin(join);
+            return leftRows <= rightRows && isZigZagJoin(join);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/join/JoinCommute.java:96]: why change to LE?

Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/join/JoinCommute.java:96]: initial planA: T1 join T2
if T1.rows = T2.rows, we need candidate planB: T2 join T1.
and then connectivity can be used to choose which plan is better

File: fe/fe-core/src/main/java/org/apache/doris/nereids/cost/CostModelV1.java (hunk 2 above)
Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/cost/CostModelV1.java:335]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/cost/CostModelV1.java (hunk 3 above)
Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/cost/CostModelV1.java:270]: done

**Review by starocean999 — DISMISSED** [review]

File: fe/fe-core/src/main/java/org/apache/doris/nereids/cost/CostModelV1.java [hunk 5]
```diff
@@ -262,6 +265,20 @@ public Cost visitPhysicalHashJoin(
 
         double leftRowCount = probeStats.getRowCount();
         double rightRowCount = buildStats.getRowCount();
+        // if the diff of the left side and right side row count is
+        // less than this threshold, we will consider their connectivity
+        double ignorableDiffBetweenLeftAndRightThreshold = 10.0;
+        if (Math.abs(leftRowCount - rightRowCount) < ignorableDiffBetweenLeftAndRightThreshold
```
Comment (xzj7019) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/cost/CostModelV1.java]: better to restrict this corn case to left row count equals to right count

Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/cost/CostModelV1.java]: done

**Review by morrySnow — APPROVED** [review]

**Review by xzj7019 — APPROVED** [review]

**Pull request merged**